cmd/zpool/zpool_main.c

```diff
@@ -2427,21 +2427,37 @@ zpool_do_import(int argc, char **argv)
 	boolean_t do_rewind = B_FALSE;
 	boolean_t xtreme_rewind = B_FALSE;
 	boolean_t do_scan = B_FALSE;
+	boolean_t do_path = B_FALSE;
 	uint64_t pool_state, txg = -1ULL;
 	char *cachefile = NULL;
 	importargs_t idata = { 0 };
 	char *endptr;
 
 	/* check options */
-	while ((c = getopt(argc, argv, ":aCc:d:DEfFmnNo:R:stT:VX")) != -1) {
+	while ((c = getopt(argc, argv, ":aCc:P:d:DEfFmnNo:R:stT:VX")) != -1) {
 		switch (c) {
 		case 'a':
 			do_all = B_TRUE;
 			break;
 		case 'c':
 			cachefile = optarg;
 			break;
+		case 'P':
+			if (searchdirs != NULL && do_path == B_FALSE) {
+				(void) fprintf(stderr,
+				    gettext("cannot mix -d and -P\n"));
+				err = 1;
+				goto error;
+			}
+			do_path = B_TRUE;
+			/* fallthrough */
 		case 'd':
+			if (c == 'd' && do_path) {
+				(void) fprintf(stderr,
+				    gettext("cannot mix -d and -P\n"));
+				err = 1;
+				goto error;
+			}
```

Comment:
This is clever exploiting the fallthrough but how about instead add a static `searchdir_add` helper function and move the alloc/realloc logic in to there.